phoenix-core/src/main/java/org/apache/phoenix/compile/OrderPreservingTracker.java

```diff
@@ -9,29 +9,37 @@
  */
 package org.apache.phoenix.compile;
 
+import java.sql.SQLException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.HashMap;
 import java.util.Iterator;
+import java.util.LinkedList;
 import java.util.List;
+import java.util.Map;
 
+import org.apache.hadoop.hbase.util.Pair;
 import org.apache.phoenix.compile.GroupByCompiler.GroupBy;
-import org.apache.phoenix.execute.TupleProjector;
+import org.apache.phoenix.compile.OrderByCompiler.OrderBy;
 import org.apache.phoenix.expression.CoerceExpression;
 import org.apache.phoenix.expression.Expression;
 import org.apache.phoenix.expression.KeyValueColumnExpression;
 import org.apache.phoenix.expression.LiteralExpression;
+import org.apache.phoenix.expression.OrderByExpression;
 import org.apache.phoenix.expression.ProjectedColumnExpression;
 import org.apache.phoenix.expression.RowKeyColumnExpression;
 import org.apache.phoenix.expression.RowValueConstructorExpression;
 import org.apache.phoenix.expression.function.FunctionExpression.OrderPreserving;
 import org.apache.phoenix.expression.function.ScalarFunction;
 import org.apache.phoenix.expression.visitor.StatelessTraverseAllExpressionVisitor;
 import org.apache.phoenix.expression.visitor.StatelessTraverseNoExpressionVisitor;
+import org.apache.phoenix.jdbc.PhoenixConnection;
 import org.apache.phoenix.schema.PTable;
-import org.apache.phoenix.schema.SortOrder;
+import org.apache.phoenix.schema.TableRef;
 import org.apache.phoenix.util.ExpressionUtil;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Iterators;
 import com.google.common.collect.Lists;
 
```

Comment:
Can you explain the changes you had to make to this class, its a little hard to follow all the changes?

Reply:
1. I added a getInputOrderBys method to determine the input OrderBys by the combination of innerQueryPlan's output OrderBys ， GroupBy of current QueryPlan or the rowKeyColumns of current table.

2. Because the innerQueryPlan may have multiple output OrderBys(for `SortMergeJoinPlan`), so I extracted many stateful memeber variables in `OrderPreservingTracker`(such as `orderPreservingTrackInfos`, `isOrderPreserving` and `isReverse` etc.) to a new inner class `TrackOrderByContext`, `TrackOrderByContext` is used to track if a single Input OrderBy matchs
the target OrderByExpressions in `OrderPreservingTracker.isOrderPreserving()`, once we found a `TrackOrderByContext`  satisfied `isOrderPreserving()`,
`OrderPreservingTracker.selectedTrackOrderByContext` is set to this `TrackOrderByContext`, then we can use this `TrackOrderByContext` to implement the `OrderPreservingTracker.getOrderPreservingTrackInfos` and `OrderPreservingTracker. isReverse` etc.